### PR TITLE
[8.0][IMP] Modifications in document approval.

### PR DESCRIPTION
**Pull request by luistorresm**
[HU #1050](https://www.vauxoo.com/web#id=1050&view_type=form&model=user.story&menu_id=557&action=726)
CA #1, 2, 3, 4, 6

[IMP][document_page_approval]
    Added field history_state in document page with widget
    kanban_state_selection, to indicate if all the histories related
    with the document are approved.
    Added test to check the new feature.
[ADD][document_page_code]Added module that add field internal code in
    document page model.
    Added in search view the new field code
[ADD][document_page_approval_levels]
    Added module that set approvers in document.page.
    Added model approver.document.page.item with fields user_id and
    approver to allow have levels to approve an history.
    Hide button approve in history when are not approved all items in
    it.
    Added as followers all users setted in the page as committee.
    Send notification to followers when is created a new history.
    Added profiles to users in knowledge
    Added users demo
    Added tests in module

### CODE REVIEW

File: document_page_approval_levels/README.rst [hunk 1]
```diff
@@ -0,0 +1,65 @@
+
+.. image:: https://img.shields.io/badge/licence-LGPL--3-blue.svg
+    :alt: License: LGPL-3
+
+Document Page Approval Levels
+=============================
+
+This module:
+  - TO-DO
```
Comment (nhomar) [inline document_page_approval_levels/README.rst]: Do not forget finish the doc.

File: document_page_approval_levels/models/document_page.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#    planned by: Sabrina Romero <sabrina@vauxoo.com>
+############################################################################
+from openerp import models, fields, api
+
+
+class DocumentPage(models.Model):
+    """This class is use to manage Document."""
+    _inherit = 'document.page'
+
+    approver_items_id = fields.One2many(
```
Comment (nhomar) [inline document_page_approval_levels/models/document_page.py]: the field has a semantic issue:

s/approver_items_id/approver_item_ids/g

Set a proper help please to explain the meaning functionally.

File: document_page_approval_levels/models/document_page_item.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                continue
+            if item._uid != item.user_id.id:
+                raise UserError(
+                    _('Invalid Action!'),
+                    _('Only the user assigned to this item can validate it.'))
+            item.write({'approved': True})
+            msg = (
+                '<div><b>The page has been approved by %s .</b></div>' %
+                item.user_id.name)
+            self.env['mail.message'].create({
```
Comment (nhomar) [inline document_page_approval_levels/models/document_page_item.py]: Use message_post not .create directly.

Do an rgrep into sale and purchase module to understand the usage.

File: document_page_code/models/document_page.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+#    planned by: Sabrina Romero <sabrina@vauxoo.com>
+############################################################################
+from openerp import models, fields
+
+
+class DocumentPage(models.Model):
+    """This class is use to manage Document."""
+    _inherit = 'document.page'
+
+    code = fields.Char('Internal code')
```
Comment (nhomar) [inline document_page_code/models/document_page.py:18]: Left the string empty, the word Code is enought.

File: document_page_approval_levels/models/document_page.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            item_obj.create({
+                'user_id': item.user_id.id,
+                'history_id': res.id,
+            })
+        return res
+
+    @api.multi
+    def write(self, vals):
+        """Added that update approvers in histories related to this
+        document when this change."""
```
Comment (nhomar) [inline document_page_approval_levels/models/document_page.py]: I prefer a server action instead a overwritting of the write method, can you use that apporach please, ask @JulioSerna for better understanding of that approach.

File: document_page_approval_levels/models/document_page.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    def create_history(self, page_id, content):
+        """Inherit method to add news items by each approver in document."""
+        item_obj = self.env['approver.document.page.item']
+        res = super(DocumentPage, self).create_history(page_id, content)
+        for item in self.browse(page_id).approver_items_id:
+            item_obj.create({
+                'user_id': item.user_id.id,
+                'history_id': res.id,
+            })
+        return res
```
Comment (nhomar) [inline document_page_approval_levels/models/document_page.py]: Tis approach can be done with a automated action alos instead and overwriting of the method which I supose is bringing a recursion problem.

File: document_page_code/models/document_page.py (hunk 4 above)
Comment (luistorresm) [inline document_page_code/models/document_page.py:18]: Thanks @nhomar 

But I have the next requirement:
`Agregar a objeto document.page campo "code" type char, el string debe ser "Internal code" y su traducción correspondiente a español "Código interno"`

File: document_page_approval_levels/models/document_page_history.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        """Return True if all Approver Items are approved."""
+        for history in self:
+            if all([item.approved for item in history.approver_items_id]):
+                history.to_approve_ok = True
+
+    @api.multi
+    def update_approver_items(self):
+        """Refresh items by user based in page_id"""
+        for history in self:
+            users_approver = [
```
Comment (JulioSerna) [inline document_page_approval_levels/models/document_page_history.py]: what about change this by one `filtered`?

File: document_page_approval_levels/models/document_page_history.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+                item.user_id.id for item in history.approver_items_id if
+                item.approved]
+            users_page = [
```
Comment (JulioSerna) [inline document_page_approval_levels/models/document_page_history.py]: What about change to `filtered`?

File: document_page_approval/document_page_approval.py [hunk 9]
```diff
@@ -245,6 +245,17 @@ def is_approval_required(self, page):
             res = False
         return res
 
+    def _get_histories_state(self):
+        """If all document histories are in approved will return done, else
+        return blocked in the document"""
+        for document in self:
+            document.history_state = 'done'
+            no_approveds = [
```
Comment (JulioSerna) [inline document_page_approval/document_page_approval.py]: what about change to `filtered`?

File: document_page_approval_levels/models/document_page.py (hunk 5 above)
Comment (JulioSerna) [inline document_page_approval_levels/models/document_page.py]: You can create one server action with this logic and the server action will be trigger when this module is update

**Pull request merged**